source/common/network/lc_trie.h

```diff
@@ -0,0 +1,540 @@
+     * @return the tag from the prefix that encompasses the input. An empty string is returned
+     *         if no prefix in the LC Trie exists.
+     */
+    std::string getTag(const IpType& ip_address) const;
+
+  private:
+    /**
+     * Builds the Level Compresesed Trie, by first sorting the tag data, removing duplicated
+     * prefixes and invoking buildRecursively() to build the trie.
+     */
+    void build(std::vector<IpPrefix<IpType>>& tag_data) {
+      if (tag_data.empty()) {
+        return;
+      }
+
+      // LcNode uses the last 20 bits to store either the index into ip_prefixes_ or trie_.
+      // In theory, the trie_ should only need twice the amount of entries of CIDR ranges.
+      // To prevent index out of bounds issues, only support a maximum of 2^19 CIDR ranges.
+      // TODO(ccaraman): Add a test case for this.
+      if (tag_data.size() > (1 << 19)) {
+        throw EnvoyException(fmt::format("Size of the input vector({0}) for LC-Trie is larger than "
+                                         "the max size supported(2^19).",
+                                         tag_data.size()));
+      }
+
+      // TODO(ccaraman): Consider adding an optimization to short circuit building the trie, if
+      // tag_data[0].length_==0.
+      std::sort(tag_data.begin(), tag_data.end());
+      ip_prefixes_.push_back(tag_data[0]);
+
+      // Remove duplicate entries and check for nested prefixes.
+      for (size_t i = 1; i < tag_data.size(); ++i) {
+        // TODO(ccaraman): Add support for nested prefixes.
+        if (tag_data[i - 1].isPrefix(tag_data[i])) {
+          throw EnvoyException(fmt::format(
+              "LcTrie does not support nested prefixes. '{0}' is a nested prefix of '{1}'.",
+              tag_data[i].asString(), tag_data[i - 1].asString()));
+        }
+
+        if (tag_data[i - 1].compare(tag_data[i]) != 0) {
+          ip_prefixes_.push_back(tag_data[i]);
+        }
+      }
+
+      // In theory, the trie_ vector can have at most twice the number of entries in the
+      // ip_prefixes_ vector. However, due to the fill factor a buffer is added to the size of the
+      // trie_. The buffer value(2000000) is reused from the reference implementation in
+      // http://www.csc.kth.se/~snilsson/software/router/C/trie.c.
+      // TODO(ccaraman): Define a better buffer value when resizing the trie_.
+      trie_.resize(2 * ip_prefixes_.size() + 2000000);
+
+      // Build the trie_.
+      uint32_t next_free_index = 1;
+      buildRecursive(0u, 0u, ip_prefixes_.size(), 0u, next_free_index);
+
+      // The value of next_free_index is the final size of the trie_.
+      trie_.resize(next_free_index);
+    }
+
+    // Thin wrapper around computeBranch output to facilitate code readability.
+    struct ComputePair {
+      ComputePair(int branch, int prefix) : branch_(branch), prefix_(prefix) {}
+
+      uint32_t branch_;
+      uint32_t prefix_;
```

Comment:
I'll add note that this is total number of bits that have the same prefix for a range of data.